Add support for adding external members to ipagroup.

rjeffman opened the pull request:
This patch add support for adding external members to ipagroup which
have the `external` attribute set. It adds another attribute to the
module, `external_members`, which is a list of users or groups from
an external trust, to be added to the group.

This patch requires server-trust-ad to be tested, as such, the tests
have been guarded by a test block, for when such tests are available
in ansible-freeipa CI.

Fixes issue #418

abbra commented:
The test playbook uses `external_members:` while `ipagroup` defines `external_member:`. After fixing that, a test that combines external group creation and adding external members fails.

Looks like using `external` and `external_member` at the same time does not work.
```
[root@master ansible-freeipa]# ansible-playbook -vvv  -i inventory.yml test_group_external_members.yml 
ansible-playbook 2.9.14
  config file = /root/ansible-freeipa/ansible.cfg
  configured module search path = ['/root/ansible-freeipa/plugins/modules']
  ansible python module location = /usr/lib/python3.8/site-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 3.8.5 (default, Aug 12 2020, 00:00:00) [GCC 10.2.1 20200723 (Red Hat 10.2.1-1)]
Using /root/ansible-freeipa/ansible.cfg as config file
host_list declined parsing /root/ansible-freeipa/inventory.yml as it did not pass its verify_file() method
script declined parsing /root/ansible-freeipa/inventory.yml as it did not pass its verify_file() method
Parsed /root/ansible-freeipa/inventory.yml inventory source with ini plugin

PLAYBOOK: test_group_external_members.yml ********************************************************************************************************************************************************************************************************************
1 plays in test_group_external_members.yml

PLAY [find trust] ********************************************************************************************************************************************************************************************************************************************
META: ran handlers

TASK [include_tasks] *****************************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:9
included: /root/ansible-freeipa/tests/env_freeipa_facts.yml for master.ipa.test

TASK [Retrieving FreeIPA version.] ***************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/tests/env_freeipa_facts.yml:10
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225 `" && echo ansible-tmp-1602616656.1659074-117602-167999629256225="` echo /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225 `" ) && sleep 0'
<master.ipa.test> Attempting python interpreter discovery
<master.ipa.test> EXEC /bin/sh -c 'echo PLATFORM; uname; echo FOUND; command -v '"'"'/usr/bin/python'"'"'; command -v '"'"'python3.7'"'"'; command -v '"'"'python3.6'"'"'; command -v '"'"'python3.5'"'"'; command -v '"'"'python2.7'"'"'; command -v '"'"'python2.6'"'"'; command -v '"'"'/usr/libexec/platform-python'"'"'; command -v '"'"'/usr/bin/python3'"'"'; command -v '"'"'python'"'"'; echo ENDFOUND && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
Using module file /usr/lib/python3.8/site-packages/ansible/modules/commands/command.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmpg7v6b42f TO /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225/AnsiballZ_command.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225/ /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225/AnsiballZ_command.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225/AnsiballZ_command.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616656.1659074-117602-167999629256225/ > /dev/null 2>&1 && sleep 0'
[DEPRECATION WARNING]: Distribution fedora 32 on host master.ipa.test should use /usr/bin/python3, but is using /usr/bin/python for backward compatibility with prior Ansible releases. A future Ansible release will default to using the discovered 
platform python for this host. See https://docs.ansible.com/ansible/2.9/reference_appendices/interpreter_discovery.html for more information. This feature will be removed in version 2.12. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
changed: [master.ipa.test] => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": true,
    "cmd": "ipa --version | sed -n \"s/VERSION: \\([^,]*\\).*API_VERSION: \\([^,]*\\).*/\\1\\\\n\\2/p\"",
    "delta": "0:00:00.538850",
    "end": "2020-10-13 19:17:37.232364",
    "invocation": {
        "module_args": {
            "_raw_params": "ipa --version | sed -n \"s/VERSION: \\([^,]*\\).*API_VERSION: \\([^,]*\\).*/\\1\\\\n\\2/p\"",
            "_uses_shell": true,
            "argv": null,
            "chdir": null,
            "creates": null,
            "executable": null,
            "removes": null,
            "stdin": null,
            "stdin_add_newline": true,
            "strip_empty_ends": true,
            "warn": true
        }
    },
    "rc": 0,
    "start": "2020-10-13 19:17:36.693514",
    "stderr": "",
    "stderr_lines": [],
    "stdout": "4.9.0.dev202010131434+git\n2.239",
    "stdout_lines": [
        "4.9.0.dev202010131434+git",
        "2.239"
    ]
}

TASK [Set FreeIPA facts.] ************************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/tests/env_freeipa_facts.yml:15
ok: [master.ipa.test] => {
    "ansible_facts": {
        "ipa_api_version": "2.239",
        "ipa_version": "4.9.0.dev202010131434+git",
        "trust_test_is_supported": true
    },
    "changed": false
}

TASK [Add nonposix group.] ***********************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:13
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682 `" && echo ansible-tmp-1602616657.4710708-117629-250776929337682="` echo /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682 `" ) && sleep 0'
Using module file /root/ansible-freeipa/plugins/modules/ipagroup.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmpdgnomwx4 TO /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682/AnsiballZ_ipagroup.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682/ /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616657.4710708-117629-250776929337682/ > /dev/null 2>&1 && sleep 0'
changed: [master.ipa.test] => {
    "changed": true,
    "failed_when_result": false,
    "invocation": {
        "module_args": {
            "action": "group",
            "description": null,
            "external": null,
            "external_member": null,
            "gid": null,
            "group": null,
            "ipaadmin_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ipaadmin_principal": "admin",
            "membermanager_group": null,
            "membermanager_user": null,
            "name": [
                "extgroup"
            ],
            "nomembers": null,
            "nonposix": true,
            "posix": null,
            "service": null,
            "state": "present",
            "user": null
        }
    }
}

TASK [Set group to be external] ******************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:21
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388 `" && echo ansible-tmp-1602616661.0109663-117647-157427300323388="` echo /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388 `" ) && sleep 0'
Using module file /root/ansible-freeipa/plugins/modules/ipagroup.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmp0qcubpu3 TO /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388/AnsiballZ_ipagroup.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388/ /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616661.0109663-117647-157427300323388/ > /dev/null 2>&1 && sleep 0'
changed: [master.ipa.test] => {
    "changed": true,
    "failed_when_result": false,
    "invocation": {
        "module_args": {
            "action": "group",
            "description": null,
            "external": true,
            "external_member": null,
            "gid": null,
            "group": null,
            "ipaadmin_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ipaadmin_principal": "admin",
            "membermanager_group": null,
            "membermanager_user": null,
            "name": [
                "extgroup"
            ],
            "nomembers": null,
            "nonposix": null,
            "posix": null,
            "service": null,
            "state": "present",
            "user": null
        }
    }
}

TASK [Add AD users to group] *********************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:29
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124 `" && echo ansible-tmp-1602616664.4630556-117666-126799009517124="` echo /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124 `" ) && sleep 0'
Using module file /root/ansible-freeipa/plugins/modules/ipagroup.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmp4bkr2how TO /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124/AnsiballZ_ipagroup.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124/ /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616664.4630556-117666-126799009517124/ > /dev/null 2>&1 && sleep 0'
changed: [master.ipa.test] => {
    "changed": true,
    "failed_when_result": false,
    "invocation": {
        "module_args": {
            "action": "group",
            "description": null,
            "external": null,
            "external_member": [
                "WIN2016\\Domain Users"
            ],
            "gid": null,
            "group": null,
            "ipaadmin_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ipaadmin_principal": "admin",
            "membermanager_group": null,
            "membermanager_user": null,
            "name": [
                "extgroup"
            ],
            "nomembers": null,
            "nonposix": null,
            "posix": null,
            "service": null,
            "state": "present",
            "user": null
        }
    }
}

TASK [Add AD users to group, again] **************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:37
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243 `" && echo ansible-tmp-1602616667.5106688-117685-223540262401243="` echo /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243 `" ) && sleep 0'
Using module file /root/ansible-freeipa/plugins/modules/ipagroup.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmplkr0_9vm TO /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243/AnsiballZ_ipagroup.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243/ /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616667.5106688-117685-223540262401243/ > /dev/null 2>&1 && sleep 0'
ok: [master.ipa.test] => {
    "changed": false,
    "failed_when_result": false,
    "invocation": {
        "module_args": {
            "action": "group",
            "description": null,
            "external": null,
            "external_member": [
                "WIN2016\\Domain Users"
            ],
            "gid": null,
            "group": null,
            "ipaadmin_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ipaadmin_principal": "admin",
            "membermanager_group": null,
            "membermanager_user": null,
            "name": [
                "extgroup"
            ],
            "nomembers": null,
            "nonposix": null,
            "posix": null,
            "service": null,
            "state": "present",
            "user": null
        }
    }
}

TASK [Remove external group] *********************************************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:45
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750 `" && echo ansible-tmp-1602616670.428519-117705-113713392048750="` echo /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750 `" ) && sleep 0'
Using module file /root/ansible-freeipa/plugins/modules/ipagroup.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmpx9tra4_t TO /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750/AnsiballZ_ipagroup.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750/ /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616670.428519-117705-113713392048750/ > /dev/null 2>&1 && sleep 0'
changed: [master.ipa.test] => {
    "changed": true,
    "failed_when_result": false,
    "invocation": {
        "module_args": {
            "action": "group",
            "description": null,
            "external": null,
            "external_member": null,
            "gid": null,
            "group": null,
            "ipaadmin_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ipaadmin_principal": "admin",
            "membermanager_group": null,
            "membermanager_user": null,
            "name": [
                "extgroup"
            ],
            "nomembers": null,
            "nonposix": null,
            "posix": null,
            "service": null,
            "state": "absent",
            "user": null
        }
    }
}

TASK [Add nonposix, external group, with AD users.] **********************************************************************************************************************************************************************************************************
task path: /root/ansible-freeipa/test_group_external_members.yml:53
<master.ipa.test> ESTABLISH LOCAL CONNECTION FOR USER: root
<master.ipa.test> EXEC /bin/sh -c 'echo ~root && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /root/.ansible/tmp `"&& mkdir "` echo /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346 `" && echo ansible-tmp-1602616673.142275-117723-91687415479346="` echo /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346 `" ) && sleep 0'
Using module file /root/ansible-freeipa/plugins/modules/ipagroup.py
<master.ipa.test> PUT /root/.ansible/tmp/ansible-local-117596rg2b0hrl/tmp4eqp3m4_ TO /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346/AnsiballZ_ipagroup.py
<master.ipa.test> EXEC /bin/sh -c 'chmod u+x /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346/ /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c '/usr/bin/python /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346/AnsiballZ_ipagroup.py && sleep 0'
<master.ipa.test> EXEC /bin/sh -c 'rm -f -r /root/.ansible/tmp/ansible-tmp-1602616673.142275-117723-91687415479346/ > /dev/null 2>&1 && sleep 0'
fatal: [master.ipa.test]: FAILED! => {
    "changed": false,
    "failed_when_result": true,
    "invocation": {
        "module_args": {
            "action": "group",
            "description": null,
            "external": true,
            "external_member": [
                "WIN2016\\Domain Users"
            ],
            "gid": null,
            "group": null,
            "ipaadmin_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "ipaadmin_principal": "admin",
            "membermanager_group": null,
            "membermanager_user": null,
            "name": [
                "extgroup"
            ],
            "nomembers": null,
            "nonposix": true,
            "posix": null,
            "service": null,
            "state": "present",
            "user": null
        }
    },
    "msg": "Cannot add external members to a non-external group."
}

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
master.ipa.test            : ok=8    changed=5    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

```

rjeffman commented:
Thank you, @abbra! 

I just force pushed what should be a fix to this. Fixed the attribute name for tests, and updated the group representation with the possibly modified values for `external` (and `posix`), so that it should be able to create a group and add external members to it in the same playbook.

chsnell commented on README-group.md:
```diff
@@ -148,6 +148,7 @@ Variable | Description | Required
 `service` | List of service name strings assigned to this group. Only usable with IPA versions 4.7 and up. | no
 `membermanager_user` | List of member manager users assigned to this group. Only usable with IPA versions 4.8.4 and up. | no
 `membermanager_group` | List of member manager groups assigned to this group. Only usable with IPA versions 4.8.4 and up. | no
+`external_member` | List of embers of a trusted domain in DOM\\name or name@domain form. | no
```
Think there's a typo in that the 3rd word should be 'members' not 'embers'

rjeffman replied:
fixed.

abbra commented:
Thanks, @rjeffman, the tests now pass in my environment:

```
# ansible-playbook -i inventory.yml test_group_external_members.yml 

PLAY [find trust] ********************************************************************************************************************************************************************************************************************************************

TASK [include_tasks] *****************************************************************************************************************************************************************************************************************************************
included: /root/ansible-freeipa/tests/env_freeipa_facts.yml for master.ipa.test

TASK [Retrieving FreeIPA version.] ***************************************************************************************************************************************************************************************************************************
[DEPRECATION WARNING]: Distribution fedora 32 on host master.ipa.test should use /usr/bin/python3, but is using /usr/bin/python for backward compatibility with prior Ansible releases. A future Ansible release will default to using the discovered 
platform python for this host. See https://docs.ansible.com/ansible/2.9/reference_appendices/interpreter_discovery.html for more information. This feature will be removed in version 2.12. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
changed: [master.ipa.test]

TASK [Set FreeIPA facts.] ************************************************************************************************************************************************************************************************************************************
ok: [master.ipa.test]

TASK [Add nonposix group.] ***********************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Set group to be external] ******************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Add AD users to group] *********************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Add AD users to group, again] **************************************************************************************************************************************************************************************************************************
ok: [master.ipa.test]

TASK [Remove external group] *********************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Add nonposix, external group, with AD users.] **********************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Add nonposix, external group, with AD users, again.] ***************************************************************************************************************************************************************************************************
ok: [master.ipa.test]

TASK [Remove group] ******************************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Add nonposix group.] ***********************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Set group to be external, and add users.] **************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

TASK [Set group to be external, and add users, again.] *******************************************************************************************************************************************************************************************************
ok: [master.ipa.test]

TASK [Cleanup environment.] **********************************************************************************************************************************************************************************************************************************
changed: [master.ipa.test]

PLAY RECAP ***************************************************************************************************************************************************************************************************************************************************
master.ipa.test            : ok=15   changed=10   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

rjeffman commented:
Thank your for testing, @abbra! 

@chsnell if you could also test it with your playbooks that were failing it would be great. As I have little experience with AD, and (still) no setup to run it.

t-woerner commented on plugins/modules/ipagroup.py:
```diff
@@ -272,6 +284,7 @@ def main():
             membermanager_user=dict(required=False, type='list', default=None),
             membermanager_group=dict(required=False, type='list',
                                      default=None),
+            external_member=dict(required=False, type='list', default=None),
```
Please add the internal name as an alias `ipaexternalmember` - also in the docs.

rjeffman replied:
Fixed.

t-woerner commented on plugins/modules/ipagroup.py:
```diff
@@ -420,40 +443,47 @@ def main():
                         service_add, service_del = gen_add_del_lists(
                             service, res_find.get("member_service"))
 
+                        external_add, external_del = gen_add_del_lists(
```
Please use `external_member_X` here.

chsnell commented:
> Thank your for testing, @abbra!
> 
> @chsnell if you could also test it with your playbooks that were failing it would be great. As I have little experience with AD, and (still) no setup to run it.

I'm having trouble setting up my environment to run the modified module, and keep pulling the released version instead.  Any tips on how to do this testing?

rjeffman commented:
You may download the branch with the modified module (https://github.com/rjeffman/ansible-freeipa/archive/fix_ipagroup_external_members_418.zip), and create a link to the plugins directory under `${HOME}/.ansible`.

```
mkdir "${HOME}/.ansible
ln -s "./ansible-freeipa_with_fix/plugins" "${HOME}/.ansible/plugins"
```
Just remove "${HOME}/.ansible/plugins" after testing to get to a system-wide configuration. (If you are using collections on your playbook, it will not work.)

chsnell commented:
@rjeffman This does appear to work (finally got it going).  Nice job!

rjeffman commented:
Thank you. This issue will be closed when PR #420 is merged. Let me know if there is still any issue.

t-woerner commented:
LGTM
